include/net/mqtt.h

```diff
@@ -69,6 +70,7 @@ struct mqtt_ctx {
 	int sock;
 	zstream stream;
 	struct zstream_sock stream_sock;
+	struct zstream_tls stream_tls;
```

Comment:
This exactly demonstrates my concern about "mixing API abstraction levels".   It looks very confusing to be having, sock, zstream and zstream_sock handles/objects all in the same client module.  I prefer to see implementations hidden under abstractions, and consistent use of abstractions.   
For example, I would rather there be only one zstream_sock object, and zstream_sock methods used on that class of object.  It also seems the sock handle is not really needed to be exposed here, and can be accessed from the zstream_sock object for poll() as needed.

Reply:
> This exactly demonstrates my concern about "mixing API abstraction levels". It looks very confusing to be having, sock, zstream and zstream_sock handles/objects all in the same client module.

So, there're 2 issues here: a) having both "fd socket" and "stream" objects; b) having "zstream stream", "struct zstream_sock stream_sock", and "struct zstream_tls stream_tls".

For b), my response remains the same - the only reason there's "mixing in API levels" is because the code mixes different operations. The only thing really needed there is "zstream stream", that's what MQTT (and any other protocol) library should receive as a parameter. Creation (and allocation) of the stream object should be done by user code (perhaps using helper functions, because creation indeed would be almost the same for any protocol library, so it only makes sense to factor it out instead of duplicating for each of them).

"a)" is more subtle issue. It's rooted in the fact that "zstream" is user-level object, and we define all operations on it. Except for one - efficient polling. That requires OS level primitive (poll()), which works directly with a socket. This issue was discussed (well, mentioned) prior the implementation work, in https://github.com/zephyrproject-rtos/zephyr/issues/5900#issuecomment-362040477 . 2 options are given there. As can be seen, I so far implemented 1st option, where a pair of (underlying_socket, higher_level_stream) is passed and used around. 2nd option is like you mention is to add a virtual method to get an underlying "fd" from any stream. The comment above explains why I didn't start with it - because it adds overhead, whereas an option to have the (underlying_socket, higher_level_stream) pair around is obvious and straightforward (YMMV). Literally, the difference is data vs code interface. The data interface offers highest efficiency, while the code one - highest flexibility, at a price.